File: src/js/TicTacToe.js


function get(name) { return localStorage.getItem(name); }


/*
	Game object
*/
function TicTacToe() {
	// this.p1 = 'human'; 			  
	// this.p2 = 'computer';			
	this.width=3; 
	this.height=3;  // initial dimension of board
	var turn;  		// whose turn it is now --> this.p1 or this.p2 
	var started; 	// true when playing game is on
	var conversationState; 	// 
	this.history = [];	// pop in/out current board state
	this.strategySet = [
						{'name':"Win",'code':"takeWin",'tooltip':'Take a cell completing three of my stones in a row/column/diagonal','enabled':true},
						{'name':"Block Win",'code':"takeBlockWin",'tooltip':"Take a cell of the opponent's winning position", 'enabled':false},
						{'name':"Take Center",'code':"takeCenter",'tooltip':"Take the center cell",'enabled':false},
						{'name':"Take Any Corner",'code':"takeAnyCorner",'tooltip':"Take any corner",'enabled':false},
						{'name':"Take Any Side",'code':"takeAnySide",'tooltip':"Take any non-corner cells on the side.",'enabled':false},
						{'name':"Random Move",'code':"takeRandom",'tooltip':"Take any empty cell.",'enabled':true},
						{'name':"Take OppositeCorner",'code':"takeOppositeCorner",'tooltip':"Take a corner cell if its opposite corner is occupied by another player",'enabled':false}
					];

 this.turn = 'p1';
 this.started = false;    
 this.board = setupBoard(this.width, this.height);
  

	/* INIT */

    function setupBoard(col, row) {
        var b = new Array();
        for (var i = 0; i < row; i++) {
            b.push(new Array(col));
        }
        return b;
    }



	/* BASIC FUNCTIONS */

    this.getTurn = function () {   	return this.turn			};
    this.historyPush = function(b,loc,t) {
    	this.history.push({'board':this.cloneBoard(b), 'loc':loc, 'turn':this.turn});
    }

	
	this.flip = function(player) {
		if (player=="p1") return "p2"
		else return "p1"
	}
	this.getBoard = function () {     return this.board;	  }
	this.cloneBoard = function(b) {
		var nb = setupBoard(this.width,this.height);
		for(var i=0;i<this.width;i++) {
			for (var j=0;j<this.height;j++) {
				nb[i][j] = b[i][j];
			}
		}
		return nb;
	}
	
	this.historyPush(this.board,undefined,undefined); // store current board and turn in history array        
 
 	this.updateRule = function() {
 		var user_id = $("#user_id").val();
 		var game_id = 'tictactoe';
 		var strategy = JSON.stringify(this.getEnabledStrategy());
 		$.get('updateRule.php',
 			{ 	user_id: user_id, 
 				game_id: game_id,
 				strategy: strategy 
 			}, function(response) {
 				alert(response);
 			}	
 		);
 	}
 
 	this.resumeAt = function(step) {
 		if (step>=this.history.length-1) {
 			alert('history out of bound');	
 		} else {
 			nB = this.history[step];
 			this.board = nB['board'];
 			if(nB['turn']=='p1') this.turn = 'p2';
 			else this.turn = 'p1';
 			// this.turn = nB['turn'];
 			this.history = this.history.slice(0,step);
 		}
 	}
	
    this.move = function(i, j, t) {
        if (this.turn != t) {
        	console.log("player turn doesn't match!" + t + "!=" + this.turn);	
        	return;
        } 
        if (this.board[i][j] != null) {
        	console.log("cannot move on occupied cell [" + i+","+j+"]");	
        	return;    	
        }
       	if (!started)   started = true;
        this.board[i][j] = this.turn;    
        this.historyPush(this.board,[i,j],t); // store current board and turn in history array        
        this.turn = this.flip(this.turn);
        return this.board[i][j];
    }

    
    this.isFull = function() {
        for (var i = 0; i < 3; i++)
			for (var j =0; j<3;j++)
			    if (this.board[i][j] == null)
			        return false;
        return true;
    }
    
	this.checkCol =function(x) {
        if (this.board[x][0] == this.board[x][1] && (this.board[x][0] == this.board[x][2]) && (this.board[x][0] != null)) 
        	return this.board[x][0];
        else return false;
    }
    
    this.checkRow = function(y) {
        if (this.board[0][y] == this.board[1][y] && (this.board[0][y] == this.board[2][y]) && (this.board[0][y] != null))
        	return this.board[0][y];
        else return false;
    }
    
    this.checkDiag1 = function()  {
        if (this.board[0][0] != null && this.board[0][0] == this.board[1][1] && this.board[0][0] == this.board[2][2]) 
        	return this.board[0][0];
        else return false;
        
    }
    
    this.checkDiag2 = function() {
        if (this.board[2][0] != null && this.board[1][1] == this.board[2][0] && this.board[0][2] == this.board[2][0]) 
        	return this.board[2][0];
        else return false;
    }
    
    this.checkForWinner = function() {
        /* Check the top row */
        var result = false;
        for (var i=0; i < 3; i++) {
            result = this.checkRow(i);
            if (result != false)
                return result;
            result = this.checkCol(i);
            if (result != false)
                return result;
        }
        result = this.checkDiag1();
        if (result != false)
            return result;
        result = this.checkDiag2();
        if (result != false)
            return result;
        return false;
    }
    
	this.analyzeMove = function(brd,player,userMoveLoc) {
		// check which strategies are matching with given brd,turn,loc
		matchingStrategy = [];
		for (key in this.strategySet) {
			st = this.strategySet[key];
			// console.log(st);
			//console("ana" + brd,player);
			result = this[st.code](brd,player);
			if (true /*result['success']==true*/) {
				for(i in result.loc) {
					l = result.loc[i];
					if(userMoveLoc[0]==l[0] && userMoveLoc[1]==l[1]) {
						matchingStrategy.push({'name':st.name,'code':st.code});
					}
					
				}				
			}
		}
		// console.log(matchingStrategy);
		return matchingStrategy;
	}
    this.restart = function() {
        this.turn = 'p1';
        this.board = setupBoard(this.width, this.height);
        this.history = [];
        this.historyPush(this.board,undefined,undefined); // store current board and turn in history array        
        this.started=false;
    }
    
    this.getStatus = function() {
        if (!this.started)
            return "New Game, player 1 to move";
        else if (this.turn == 'p1')
            return "player 1 to move";
        else
            return "player 2 to move";
        
    }
    
    this.getEnabledStrategy = function() {
    	enabledStrategy = [];
    	for (i in this.strategySet) {
    		st = this.strategySet[i];  // select one strategy one-by-one
    		if(st['enabled']==false) continue;  // if the strategy is not enabled, pass
    		else enabledStrategy.push(st);
    	}
    	return enabledStrategy;
    }
    
    this.getEnabledStrategy_name = function() {
    	enabledStrategy = [];
    	for (i in this.strategySet) {
    		st = this.strategySet[i];  // select one strategy one-by-one
    		if(st['enabled']==false) continue;  // if the strategy is not enabled, pass
    		else enabledStrategy.push(st['code']);
    	}
    	return enabledStrategy;
    }
    
    this.enableStrategy = function(code) {
    	for (i in this.strategySet) {
    		st = this.strategySet[i];
    		if(st['code']==code) {
    			if(st['enabled']) {
    				return false;
    			} else {
    				st['enabled']=true;
    				return true;	
    			}
    		}
    	}
    	return 
    }
    
    this.checkStrategyEnabled = function(code) {
    	for (i in this.strategySet) {
    		if(st['code']==code) {
    			return st['enabled'];
    		}
    	}
    }
    
    this.changeOrder = function(nameList) {
    	newStrategySet = [];  
    	// console.log(nameList);
    	for (ni in nameList) {
    		name = nameList[ni];
    		for(i in this.strategySet) {
    			st=this.strategySet[i];
    			if(st['name']==name) {
    				newStrategySet.push(st);
    			}
    		}
    	}
    	console.log(newStrategySet);
    	// add strategies not in the codeList (new strategy order)
    	for(si in this.strategySet) {
    		st = this.strategySet[si];
    		alreadyAdded = false;
    		for (ni in nameList) {
    			name = nameList[ni];
    			if (name==st['name']) alreadyAdded = true;
    		}
    		if (alreadyAdded==false) newStrategySet.push(st);
    	}
    	this.strategySet = newStrategySet;
    }
    
    this.findBestStrategy = function(brd,player,strategy) {
    	if (this.isFull()) {
    		return {'message':"Tie Game",'locList':undefined};
    	}
    	for (key in strategy) {
    		st = strategy[key];  // select one strategy one-by-one
    		if(st['enabled']==false) continue;  // if the strategy is not enabled, pass
    		result = this[st.code](brd,player); // try the strategy on current board
    		if (result['success']==true)  {
    			bestStrategy = {'message':st.name,'locList':result.loc};
    			return bestStrategy;	// return the first strategy matched. d'nt try the rest
    		}
    	}
		bestStrategy = {'message':'no strategy found','locList':undefined};
		return bestStrategy;
    }
    
    
    
    
    
	/* STRATEGIES */
    this.takeWin = function(brd,player) {
		var result = new Object();		
		result['success']=false;
		result['loc'] = [];		// one strategy can generate multiple locations
        // Check diagonals for win
        var combos = [
            [[0,0],[1,1],[2,2]],
            [[0,2],[1,1], [2,0]], 
            
            [[0,0], [0,1], [0,2]], 
            [[1,0], [1,1], [1,2]], 
            [[2,0], [2,1], [2,2]], 
            
            [[0,0], [1,0], [2,0]], 
            [[0,1], [1,1], [2,1]],
            [[0,2], [1,2], [2,2]],         
        ];
        
        for (var  i in combos) {
            var positions = combos[i];
            var open = null;
            var count = 0;
            for (var pos in positions) {
                var move = positions[pos];
                var x = move[0];
                var y = move[1];
                //console.log(brd, player);
                if (brd[x][y] == player)
                    count++;
                else if (!brd[x][y]) 
                    open = move;
            }
            
            if (count == 2 && open) {
				result['success']=true;
				result['loc'].push(open);	
				// alert(result['loc']);
				//return result;
            }
        }
        
		return result;
        
    }
    this.takeBlockWin = function(brd,player) {
		var result = new Object();
		result['success']=false;
		result['loc'] = [];		
        var origTurn = player
        // Flip turn
        player =  this.flip(player);;
        var willWin = this.takeWin(brd,player);
        // Restore original turn
        player = origTurn;
		if (willWin['success']) {
			return willWin;		
		} else {
			// no winning condition for opponent
			result['success']=false;
			return result;
		}		
    }
    this.takeCenter = function(brd,player) {
    	// alert("take Center");
		var result = new Object();
		result['success']=false;  
		result['loc'] = [];	
		// console.log('check takecenter' + brd[1][1]);
        if (!brd[1][1]) {
			result['success']=true;
			result['loc'].push([1,1]);	
			// alert(result['loc']);
			return result;
        } else {
        	// alert('center occupied');
			result['success']=false;	
			return result;
		}
    }
	this.takeAnyCorner = function(brd,player) {
		var result = new Object();	
		result['success']=false;
		result['loc'] = [];	
		var moves = [[0,0], [2,2], [2,0], [0, 2]];
		for (var i=0;i<moves.length; i++) {
           var move=moves[i];
           var x = move[0];
           var y = move[1];
           
           if (!brd[x][y]) {
               // this.move(x,y);
				result['success']=true;
				result['loc'].push([x, y]);	
				//return result;
           }
		}
		// result['success']=false;
		return result;
    }
    this.takeAnySide = function(brd,player) {
        var result = new Object();
        result['success']=false;
		result['loc'] = [];	
        var moves = [[0,1], [1,0], [1,2], [2,1]];
        for (var i=0; i<moves.length;i++) {
            var pt = moves[i];
            if (!brd[pt[0]][pt[1]]) {
                result['success']=true;
				result['loc'].push([pt[0],pt[1]]);	
            }
        }
        return result;
    }
    this.takeRandom = function(brd,player) {	
		var result = new Object();	
		result['success']=false;
		result['loc'] = [];
		for(var i=0;i<this.width;i++) {
			for (var j=0;j<this.height;j++) {
				if(!brd[i][j]) {
					result['success']=true;
					result['loc'].push([i,j]);
				}
			}
		}	
		return result;
    }
	this.stopL = function (brd,player) {
	    // block corner trap strategy : two adj. sides are occupied by opponent, 
	    //								and in-between corner is empty.
	    // 								if I don't take the corner, opp. will take it. 
    	var result = new Object();	
 		result['success']=false;
		result['loc'] = [];	
        var oppTurn =  this.flip(player);
        if (!brd[0][0] && brd[0][1] == oppTurn && brd[1][0] == oppTurn)
			result['success']=true;
			result['loc'].push([0,0]);	
			// return result;		
        if (!brd[0][2] && brd[0][1] == oppTurn && brd[1][2] == oppTurn)
			result['success']=true;
			result['loc'].push([0,2]);	
			// return result;	
        if (!brd[2][0] && brd[1][0] == oppTurn && brd[2][1] == oppTurn)
			result['success']=true;
			result['loc'].push([2,0]);	
			// return result;
        if (!brd[2][2] && (brd[1][2] == oppTurn) && (brd[2][1] == oppTurn))
			result['success']=true;
			result['loc'].push([2,2]);	
			// return result;
        return result;
    }
    this.takeOppositeCorner = function(brd,player) {
		var result = new Object();	
   		result['success']=false;
		result['loc'] = [];	
        var pairs = [ [[0,0], [2,2]], [[0,2], [2,0]] ];  // for each pair of opposing corners
        for (var i=0;i<pairs.length; i++) {		
              var pair=pairs[i];	
              var pt1 = pair[0];	// x
              var pt2 = pair[1];	// y
              var ptOccupied1 = brd[pt1[0]][pt1[1]]; 
              var ptOccupied2 = brd[pt2[0]][pt2[1]];
              
              // if one corner is occupied, take another one. 
              if (ptOccupied1 && !ptOccupied2) { 
				result['success']=true;
				result['loc'].push(pt2);	
				// return result;
              }
              else if (ptOccupied2 && !ptOccupied1) {
				result['success']=true;
				result['loc'].push(pt1);	
				// return result;
              }
                
		}
		// result['success']=false;
		return result;
    }
}

// //
// var TTT_WIN = "Win";
// var TTT_CENTER = "Take Center";
// var TTT_CORNER = "Take any corner";
// var TTT_OPP_CORNER = "Take opposite corner";
// var TTT_BLOCK_WIN = "Block Win";
// var TTT_TAKE_SIDE = "Take any side";
// var TTT_RANDOM = "Random Move";
// var TTT_BLOCK_CORNER_TRAP = "Block Corner Trap";

File: src/TicTacToe.py
'''
Created on Dec 6, 2011
    merging TicTacToeMatch and TicTacToeTrainer into one
@author: reflect9
'''

from google.appengine.ext import db
import sys, copy, json
from random import choice
from django.utils import simplejson
import datastore

''' TRAINER FUNCTIONS '''
# for adding user-created custom rule to Rule table
def addCustomRule(ruleBoard, title, desc, author, translationInvariant,
                    flipping, rowPermutation, columnPermutation, rotation):
    def addRule(container, toBeAdded):
        if toBeAdded not in container:
            container.append(toBeAdded)
    def flip(ruleBoard, horizontally, vertically):
        flipped = []
        for i in range(len(ruleBoard)):
            flipped.append([0]*len(ruleBoard[0]))
        for i in range(len(ruleBoard)):
            for j in range(len(ruleBoard[0])):
                xIndex = i
                yIndex = j
                if horizontally:
                    xIndex = len(ruleBoard)-i-1
                if vertically:
                    yIndex = len(ruleBoard[0])-j-1
                flipped[xIndex][yIndex] = ruleBoard[i][j]
        return flipped
    def rotate(ruleBoard):
        # rotated will start with transposed dimensions
        rotated = []
        for i in range(len(ruleBoard[0])):
            rotated.append([0]*len(ruleBoard))
        # now define the values appropriately
        for i in range(len(ruleBoard)):
            for j in range(len(ruleBoard[0])):
                xIndex = j
                yIndex = len(ruleBoard)-i-1
                rotated[xIndex][yIndex] = ruleBoard[i][j]
        return rotated
    def colPermute(ruleBoard,offset):
        permuted = []
        for i in range(len(ruleBoard)):
            permuted.append([0]*len(ruleBoard[0]))
        print permuted
        print ruleBoard
        for i in range(len(ruleBoard)):
            for j in range(len(ruleBoard[0])):
                xIndex = (i + offset) % len(ruleBoard)
                permuted[xIndex][j] = ruleBoard[i][j]
                print (i,",",j,": ",permuted)
        return permuted
    def rowPermute(ruleBoard,offset):
        permuted = []
        for i in range(len(ruleBoard)):
            permuted.append([0]*len(ruleBoard[0]))
        for i in range(len(ruleBoard)):
            for j in range(len(ruleBoard[0])):
                yIndex = (i + offset) % len(ruleBoard[0])
                permuted[i][yIndex] = ruleBoard[i][j]
        return permuted
    def minimize(ruleBoard):
        hMin = len(ruleBoard)
        hMax = 0
        vMin = len(ruleBoard[0])
        vMax = 0
        for i in range(len(ruleBoard)):
            for j in range(len(ruleBoard[0])):
                if ruleBoard[i][j] != CellType.IGNORE:
                    hMin = min(hMin,i)
                    hMax = max(hMax,i)
                    vMin = min(vMin,j)
                    vMax = max(vMax,j)
        width = hMax-hMin+1
        height = vMax-vMin+1

        newBoard = []
        for i in range(hMin,hMin+width):
            newBoard.append([])
            for j in range(vMin,vMin+height):
                newBoard[i].append(ruleBoard[i][j])
        return newBoard
    # Here is the start of the actual processing function
    if translationInvariant:
        ruleBoard = minimize(ruleBoard)
    boardList = []
    boardList.append(ruleBoard)
    if flipping:
        for i in range(len(boardList)):
            print boardList
            addRule(boardList,flip(boardList[i],True,False))
            print boardList
            addRule(boardList,flip(boardList[i],False,True))
            print boardList
            addRule(boardList,flip(boardList[i],True,True))
            print boardList
    if rowPermutation:
        for i in range(len(boardList)):
            for j in range(len(boardList[i][0])):
                addRule(boardList,rowPermute(boardList[i],j))
    if columnPermutation:
        for i in range(len(boardList)):
            for j in range(len(boardList[i])):
                addRule(boardList,colPermute(boardList[i],j))
    print >>sys.stderr, boardList
    if rotation:
        for i in range(len(boardList)):
            rotated = rotate(boardList[i])
            addRule(boardList,rotated)
            addRule(boardList,flip(boardList[i],True,True))
            addRule(boardList,flip(rotated,True,True))

    print boardList.__repr__()
    newRule = datastore.Rule(title=title,code='',description=desc,author=author,game='tictactoe',type='board definition',definition=boardList.__repr__()).put()
    return newRule

# for adding a built-in rule to user's AI by title
def activateBuiltInRule(userID,ruleKeyString):
#    rule = db.GqlQuery("SELECT * FROM Rule WHERE title=:1 AND rule_type=:2",ruleTitle,'built-in').get()
    ai = datastore.getAI(userID,'tictactoe')
    ruleKey = db.Key(ruleKeyString)
    return ai.addRule(ruleKey)
def activateBuiltInRuleByTitle(userID,ruleTitle):
    rule = db.GqlQuery("SELECT * FROM Rule WHERE title=:1 AND rule_type=:2",ruleTitle,'built-in').get()
    ai = datastore.getAI(userID,'tictactoe')
    ruleKey = rule.key()
    return ai.addRule(ruleKey)

''' MATCH FUNCTIONS'''
# run a single match and return result
def runMatch(p1,p2,startTurn):
    strategy = {};  history=[]; counter = 0; result = ""
    strategy[p1] = datastore.getUserRule(p1,'tictactoe')  
    strategy[p2] = datastore.getUserRule(p2,'tictactoe')  
    turn = startTurn
    board = [[0 for i in range(3)] for j in range(3)]
    def makeMove(board,x,y,turn):
        if board[x][y] != 0:
            print >>sys.stderr,"cannot move on occipied cell ["+str(x)+","+str(y)+"]"
            return
        board[x][y]=turn
        return board
    while True:
        opponent = p2 if turn==p1 else p1
        nextMove = findBestStrategy(board,turn,opponent)
        if nextMove['message']=="Tie Game":
            history.append({'board':copy.deepcopy(board),'loc':None,'turn':turn,'message':nextMove['message']})
            result = "Tie Game"
            return {'history':history, 'winner': result}       
        else: # now it will select one from all the moves of the best strategy
            selectedLoc = choice(nextMove['locList'])  # randomly select one location from list
            board = makeMove(board,selectedLoc[0], selectedLoc[1],turn); # update board
            history.append({'board':copy.deepcopy(board),'loc':selectedLoc,'turn':turn,'message':nextMove['message']})
            winner = checkWinner(board);
            if winner:
                result = winner
                return {'history':history, 'winner': result}       
            turn = p2 if turn==p1 else p1
        counter = counter+1
        if counter>9: break            
    return {'history':history, 'winner': result}    
# run multiple matches
def runMatches(p1,p2,numberOfMatches):
    matches = []
    firstTurn = p1
    for i in range(0,numberOfMatches):
        if i<(numberOfMatches/2):    firstTurn = p1
        else:       firstTurn = p2
        match = runMatch(p1,p2,firstTurn)
        matches.append(match)
    p1_AI = datastore.getUserRule(p1, 'tictactoe')
    p2_AI = datastore.getUserRule(p2, 'tictactoe')
    result = {}
    result['players'] = {"p1":p1, "p2":p2}
    result['AI'] = {p1:p1_AI, p2:p2_AI}
    result['matches'] = matches 
    return result

''' GENERAL UTILITY FUNCTIONS '''
def findBestStrategy(board,turn):  
    if isFull(board):       return {'message':"Tie Game",'locList':None}
    # retrieve player's rules from DB here
    userID = turn.replace("_AI","")
    rules = datastore.getUserRuleDict(userID,'tictactoe')      
    for rule in rules:
        if (rule['rule_type']=="built-in"):
            result = eval(rule['definition'])(board,turn)
        elif (rule['rule_type']=="board definition"):
            result = evaluateCreatedStrategy(rule['definition'],board,turn)
        if result['success']:
            return {'message':rule['title'],'rule':rule,'userRules':rules,'locList':result['loc']}
    return {'message':"no matching strategy found",'userRules':rules,'locList':None}

def findMatchingStrategy(board, player, userLoc):
    matchingStrategy = [];
    builtInRules = datastore.getBuiltInRule('tictactoe')
    userLoc = json.loads(userLoc)
    userRules = datastore.getAI(player,'tictactoe').data
    for rule in builtInRules:
        if rule['key'] in userRules:
            rule['enabled']=True # PREVIOUSLY st['code']
        else:  rule['enabled']=False
#        p1p2 = 'p1' if self.turn==self.p1 else 'p2'
#        print >>sys.stderr, "board : " + json.dumps(self.board) + p1p2
        if (rule['rule_type']=="built-in"):
            result = eval(rule['definition'])(board,player)
        elif (rule['rule_type']=="board definition"):
            result = evaluateCreatedStrategy(eval(rule['definition']), board, player)
        if result['success'] and result['loc']!=None:
            for loc in result['loc']:
                if int(userLoc[0])==loc[0] and int(userLoc[1])==loc[1]:
                    matchingStrategy.append(rule)
                else:
                    print >>sys.stderr, str(userLoc) + "__" + str(loc)
    return matchingStrategy

''' BASIC FUNCTIONS FOR CHECKING BOARD STATUS '''
def isFull(board):
    for row in board:  
        for cell in row:
            if cell==0: return False
    return True
def checkCol(board,x):
    if board[x][0]==board[x][1] and board[x][1]==board[x][2] and board[x][0]!=0:
        return board[x][0]
    else: return False
def checkRow(board,y):
    if board[0][y]==board[1][y] and board[1][y]==board[2][y] and board[0][y]!=0:
        return board[0][y]
    else: return False
def checkDiag1(board):
    if board[0][0]!=0 and board[0][0]==board[1][1] and board[1][1]==board[2][2]:
        return board[0][0]
    else: return False
def checkDiag2(board):
    if board[2][0]!=0 and board[2][0]==board[1][1] and board[0][2]==board[1][1]:
        return board[2][0]
    else: return False      
def checkWinner(board):
    for i in range(0,3):
        if checkRow(board,i) != False: 
            return checkRow(board,i)
        if checkCol(board,i) != False: 
            return checkCol(board,i)
    if checkDiag1(board)!=False: 
        return checkDiag1(board)
    if checkDiag2(board)!=False:
        return checkDiag2(board) 
    return False


''' BUILT-IN STRATEGIES'''
def takeWin(board,player):
    combos = [
              [[0,0],[1,1],[2,2]],
              [[0,2],[1,1],[2,0]],
              
              [[0,0],[0,1],[0,2]],
              [[1,0],[1,1],[1,2]],
              [[2,0],[2,1],[2,2]],
              
              [[0,0],[1,0],[2,0]],
              [[0,1],[1,1],[2,1]],
              [[0,2],[1,2],[2,2]]                  
              ]
    for case in combos:
        countEmptyCell = 0
        countPlayerCell = 0
        for pos in case:
            if board[pos[0]][pos[1]]==player:
                countPlayerCell = countPlayerCell+1
            if board[pos[0]][pos[1]]==0:
                emptyCell = pos
                countEmptyCell = countEmptyCell+1
        if countEmptyCell==1 and countPlayerCell==2:
            return {'success':True, 'loc':[emptyCell]}
    return {'success':False,'loc':None}
def takeBlockWin(board,player):
    combos = [
              [[0,0],[1,1],[2,2]],
              [[0,2],[1,1],[2,0]],
              
              [[0,0],[0,1],[0,2]],
              [[1,0],[1,1],[1,2]],
              [[2,0],[2,1],[2,2]],
              
              [[0,0],[1,0],[2,0]],
              [[0,1],[1,1],[2,1]],
              [[0,2],[1,2],[2,2]]                  
              ]
    for case in combos:
        countEmptyCell = 0
        countPlayerCell = 0
        for pos in case:
            # THE ONLY DIFFERENCE FROM takeWin IS THE LINE BELOW 
            if board[pos[0]][pos[1]]!=player and board[pos[0]][pos[1]]!=0:  
                countPlayerCell = countPlayerCell+1
            if board[pos[0]][pos[1]]==0:
                emptyCell = pos
                countEmptyCell = countEmptyCell+1
        if countEmptyCell==1 and countPlayerCell==2:
            return {'success':True, 'loc':[emptyCell]}
    return {'success':False,'loc':None}
def takeCenter(board,player):
    if board[1][1]==0:
        return {'success':True, 'loc':[[1,1]]}
    return {'success':False}
def takeAnyCorner(board,player):
    combos = [[0,0],[2,2],[2,0],[0,2]]
    possibleMoves = []
    for case in combos:
        if board[case[0]][case[1]]==0:
            possibleMoves.append(case)
    if len(possibleMoves)>0:
        return {'success':True, 'loc':possibleMoves}
    else:
        return {'success':False, 'loc':None}
def takeAnySide(board,player):
    combos = [[0,1],[1,0],[1,2],[2,1]]
    possibleMoves = []
    for case in combos:
        if board[case[0]][case[1]]==0:
            possibleMoves.append(case)
    if len(possibleMoves)>0:
        return {'success':True, 'loc':possibleMoves}
    else:
        return {'success':False, 'loc':None}
def takeRandom(board,player):
    possibleMoves = []
    for x in range(0,3):
        for y in range(0,3):
            if board[x][y]==0:  possibleMoves.append([x,y])
    if len(possibleMoves)>0:
        return {'success':True, 'loc':possibleMoves}
    else:
        return {'success':False, 'loc':None}
def takeOppositeCorner(board,player):
    combos = [[[0,0],[2,2]],[[0,2],[2,0]]]
    for corners in combos:
        c1 = corners[0]
        c2 = corners[1]
        if board[c1[0]][c1[1]]!=0 and board[c2[0]][c2[1]]==0:
            return {'success':True, 'loc':[c2]}
        if board[c2[0]][c1[1]]!=0 and board[c1[0]][c2[1]]==0:
            return {'success':True, 'loc':[c1]}      
    return {'success':False}
def evaluateCreatedStrategy(ruleBoardList, board, player):
    result = {'success':True,'loc':[]}
    for row in board:
        for cell in row:
            if cell!=0 and cell!=player:
                opponent = cell
                break
    pushFlag = False
    ruleBoardList = eval(ruleBoardList)
    for ruleBoard in ruleBoardList:
        width = len(ruleBoard)
        height = len(ruleBoard[0])
        for i in range(len(board)-width+1):
            for j in range(len(board[0])-height+1):
                for k in range(i,i+width):
                    for l in range(j,j+height):
                        tile = ruleBoard[k-i][l-j]
                        if (tile == CellType.IGNORE):
                            continue
                        if (tile == CellType.P1 and board[k][l] != player):
                            result['success'] = False
                        elif (tile == CellType.P2 and board[k][l] != opponent):
                            result['success'] = False
                        elif (tile == CellType.EMPTY and board[k][l] != CellType.EMPTY):
                            result['success'] = False
                        elif (tile == CellType.SELECTED and board[k][l] == CellType.EMPTY):
                            result['loc'].append([k,l])
                            pushFlag = True
                if pushFlag and not result['success']:
                    result['loc'].pop()
                pushFlag = False
                result['success'] = True
    if len(result['loc']) == 0:
        result['success'] = False
    return result
class CellType:
    EMPTY = 0
    P1 = 1
    P2 = 2
    SELECTED = 3
    IGNORE = 4
